// Generated from departure_orchestration_service.ifex by covesa-ifex-core
// DO NOT EDIT - regenerate with: ./generate_proto.sh
syntax = "proto3";



package swdv.departure_orchestration_service;

// Namespace: departure
// Departure scheduling and preparation orchestration

// Types of preparation steps
enum preparation_step_type_t {
  CLIMATE_START = 0;
  DEFROST_START = 1;
  SEAT_HEATING_START = 2;
  BEVERAGE_START = 3;
  BATTERY_PRECONDITIONING = 4;
  ROUTE_CALCULATION = 5;
}

// Status of departure schedule
enum schedule_status_t {
  SCHEDULED = 0;
  PREPARING = 1;
  READY = 2;
  COMPLETED = 3;
  CANCELLED = 4;
  ERROR = 5;
}

// Comfort preferences for departure
message comfort_preferences_t {
  float cabin_temperature_celsius = 1;
  sint32 seat_temperature_level = 2;
  bool steering_wheel_heating = 3;
}

// Beverage preparation request
message beverage_request_t {
  bool prepare_beverage = 1;
  string beverage_type = 2;
  uint32 strength = 3;
}

// Complete departure schedule configuration
message departure_schedule_t {
  sint64 departure_time = 1;
  comfort_preferences_t comfort_prefs = 2;
  beverage_request_t beverage_request = 3;
  bool auto_defrost = 4;
  bool precondition_battery = 5;
  string calculate_route_to = 6;
}

// Individual preparation step in timeline
message preparation_step_t {
  preparation_step_type_t step_type = 1;
  sint64 scheduled_start_time = 2;
  uint32 duration_seconds = 3;
  string description = 4;
}

// Detailed schedule status information
message schedule_status_info_t {
  string schedule_id = 1;
  schedule_status_t status = 2;
  sint64 departure_time = 3;
  sint64 preparation_start_time = 4;
  string current_step = 5;
  uint32 completion_percent = 6;
}

// Method: schedule_departure
// Create a new departure schedule
message schedule_departure_request {
  departure_schedule_t schedule = 1;
}

message schedule_departure_response {
  string schedule_id = 1;
  bool accepted = 2;
  repeated preparation_step_t preparation_timeline = 3;
}

service schedule_departure_service {
  rpc schedule_departure(schedule_departure_request) returns (schedule_departure_response);
}

// Method: get_schedule_status
// Get status of a departure schedule
message get_schedule_status_request {
  string schedule_id = 1;
}

message get_schedule_status_response {
  schedule_status_info_t status_info = 1;
  repeated preparation_step_t active_steps = 2;
}

service get_schedule_status_service {
  rpc get_schedule_status(get_schedule_status_request) returns (get_schedule_status_response);
}

// Method: cancel_schedule
// Cancel a departure schedule
message cancel_schedule_request {
  string schedule_id = 1;
}

message cancel_schedule_response {
  bool success = 1;
  repeated string systems_to_shutdown = 2;
}

service cancel_schedule_service {
  rpc cancel_schedule(cancel_schedule_request) returns (cancel_schedule_response);
}

// Method: modify_schedule
// Modify an existing schedule
message modify_schedule_request {
  string schedule_id = 1;
  sint64 new_departure_time = 2;
  comfort_preferences_t new_comfort_prefs = 3;
}

message modify_schedule_response {
  bool success = 1;
  repeated preparation_step_t updated_timeline = 2;
}

service modify_schedule_service {
  rpc modify_schedule(modify_schedule_request) returns (modify_schedule_response);
}

// Method: get_active_schedules
// Get all active departure schedules
message get_active_schedules_request {
}

message get_active_schedules_response {
  repeated schedule_status_info_t schedules = 1;
}

service get_active_schedules_service {
  rpc get_active_schedules(get_active_schedules_request) returns (get_active_schedules_response);
}


